python/imjoy_rpc/connection/socketio_connection.py

```diff
@@ -1,37 +1,142 @@
 import uuid
+import sys
+import asyncio
 import socketio
+import logging
 from imjoy_rpc.utils import MessageEmitter, dotdict
+from imjoy_rpc.rpc import RPC
+import contextvars
+
+logging.basicConfig(stream=sys.stdout)
+logger = logging.getLogger("SocketIOConnection")
+logger.setLevel(logging.INFO)
+
+connection_id = contextvars.ContextVar("connection_id")
+
+
+class SocketIOManager:
+    def __init__(self, rpc_context):
+        self.default_config = rpc_context.default_config
+        self.clients = {}
+        self.interface = None
+        self.rpc_context = rpc_context
+        self._codecs = {}
+
+    def get_ident(self):
+        return connection_id.get(default=None)
+
+    def reset_interface(self):
+        self.set_interface({})
+
+    def set_interface(self, interface, config=None):
+        config = config or self.default_config
+        config = dotdict(config)
+        config.name = config.name or "ImJoy Plugin"
+        config.allow_execution = config.allow_execution or False
+        config.version = config.version or "0.1.0"
+        config.api_version = config.api_version or "0.2.3"
+        config.description = config.description or "[TODO: add description]"
+        config.id = config.id or str(uuid.uuid4())
+        self.default_config = config
+        self.interface = interface
+        for k in self.clients:
+            self.clients[k].rpc.set_interface(interface, self.default_config)
+
+        if config.plugin_server:
+            asyncio.ensure_future(self.sio.connect(config.plugin_server))
+
+    def register_codec(self, config):
+        assert "name" in config
+        assert "encoder" in config or "decoder" in config
+        if "type" in config:
+            for tp in list(self._codecs.keys()):
+                codec = self._codecs[tp]
+                if codec.type == config["type"] or tp == config["name"]:
+                    logger.warn("Removing duplicated codec: " + tp)
+                    del self._codecs[tp]
+
+        self._codecs[config["name"]] = dotdict(config)
+
+    def start(self, channel):
+        sio = socketio.AsyncClient()
+
+        @sio.on("join_rpc_channel")
+        def join_rpc_channel(data):
+            self._create_new_connection(sio, data["sid"])
+
+        @sio.event
+        async def connect():
+            await sio.emit("register_plugin", {"id": sio.sid}, callback=registered)
+
+        self.sio = sio
+        print("plugin started.")
+        loop.run_until_complete(sio.wait())
+
+        if url:
+            logger.info("connecting to %s", url)
+            asyncio.ensure_future(sio.connect(url))
```

Comment:
`asyncio.ensure_future` is not thread safe. We need to use a thread safe method to schedule the coroutine on the event loop from a different thread.

https://docs.python.org/3/library/asyncio-task.html#scheduling-from-other-threads

Reply:
I guess the connection logic will only run in a single thread. It will require more work to support multi-threading, we will also need to change `rpc.py`.

Is there drawback if we use function ended with `_threadsafe` when we call them in a single thread? Slower?

Reply:
When the event loop has started there will always be at minimum two threads, the parent thread that started the event loop and the event loop thread.

Most functions in the asyncio library are only safe to use inside the event loop. That's why they are called non threadsafe.

`loop.run_until_complete` is used outside the event loop.

Reply:
There is some issue with this PR, I will fix them in the following PR.